### PR TITLE
FOUNDATIONS: Add Event

### PR DESCRIPTION
**Pull request by rcruzfreelance**

### CODE REVIEW

**Comment by rcruzfreelance** [comment]
Closes #259

File: Taarafo.Core/Services/Foundations/Events/EventService.cs [hunk 1]
```diff
@@ -27,6 +28,14 @@ public partial class EventService : IEventService
             this.loggingBroker = loggingBroker;
         }
 
+        public ValueTask<Event> AddEventAsync(Event @event) =>
+        TryCatch(async () =>
```
Comment (jasurbekyusuf) [inline Taarafo.Core/Services/Foundations/Events/EventService.cs:32]: Looks good.  Can you just indent the TryCatch

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                    Times.Once());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedEventDependencyValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertEventAsync(It.IsAny<Event>()),
+                        Times.Never);
```
Comment (jasurbekyusuf) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs:165]: Remove tab

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+                new DuplicateKeyException(randomMessage);
+
+            var alreadyExistsEventException =
+                new AlreadyExistsEventException(duplicateKeyException);
+
+            var expectedEventDependencyValidationException =
+                new EventDependencyValidationException(alreadyExistsEventException);
+
+            this.dateTimeBrokerMock.Setup(broker =>
+              broker.GetCurrentDateTimeOffset())
```
Comment (jasurbekyusuf) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs:88]: Add tabe (....)

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Validations.Add.cs [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+                    addEventTask.AsTask);
+
+            // then
+            actualEventValidationException.Should().BeEquivalentTo(
+                expectedEventValidationException);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedEventValidationException))),
+                    Times.Once);
```
Comment (jasurbekyusuf) [inline Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Validations.Add.cs:45]: Add tabe ....

File: Taarafo.Core/Services/Foundations/Events/EventService.Exceptions.cs [hunk 5]
```diff
@@ -37,6 +41,75 @@ private IQueryable<Event> TryCatch(ReturningEventsFunction returningEventsFuncti
             }
         }
 
+        private async ValueTask<Event> TryCatch(ReturningEventFunction returningEventFunction)
```
Comment (jasurbekyusuf) [inline Taarafo.Core/Services/Foundations/Events/EventService.Exceptions.cs:44]: Write ReturningEventFunction() before ReturningEventsFunction()

Comment (rcruzfreelance) [reply Taarafo.Core/Services/Foundations/Events/EventService.Exceptions.cs:23]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Validations.Add.cs [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+                    expectedEventValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertEventAsync(invalidEvent),
+                    Times.Never);
+
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```
[the remark this answers is not in the document]
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Validations.Add.cs]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Validations.Add.cs (hunk 4 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Validations.Add.cs]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs (hunk 3 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs]: done

File: Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs (hunk 2 above)
Comment (rcruzfreelance) [reply Taarafo.Core.Tests.Unit/Services/Foundations/Events/EventServiceTests.Exceptions.Add.cs]: done

File: Taarafo.Core/Services/Foundations/Events/EventService.cs (hunk 1 above)
Comment (rcruzfreelance) [reply Taarafo.Core/Services/Foundations/Events/EventService.cs]: done

**Comment by rcruzfreelance** [comment]
Please let me know if we still need something from here

File: Taarafo.Core/Models/Events/Event.cs [hunk 7]
```diff
@@ -12,12 +12,12 @@ public class Event
 		public Guid Id { get; set; }
 		public string Name { get; set; }
 		public string Description { get; set; }
-		public DateTime Date { get; set; }
+		public DateTimeOffset Date { get; set; }
```
Comment (glhays) [inline Taarafo.Core/Models/Events/Event.cs:15]: Changing the type here, will need a migration to adjust the Event table.
Like wise with the other columns.

**Review by glhays — APPROVED** [review]

**Pull request merged**